exporter/http.go

```diff
@@ -170,23 +171,44 @@ func (e *Exporter) reloadPwdFile(w http.ResponseWriter, r *http.Request) {
 }
 
 func (e *Exporter) isBasicAuthConfigured() bool {
-	return e.options.BasicAuthUsername != "" && e.options.BasicAuthPassword != ""
+	return e.options.BasicAuthUsername != "" && (e.options.BasicAuthPassword != "" || e.options.BasicAuthHashPassword != "")
 }
 
 func (e *Exporter) verifyBasicAuth(user, password string, authHeaderSet bool) error {
-
 	if !e.isBasicAuthConfigured() {
 		return nil
 	}
 
 	if !authHeaderSet {
 		return errors.New("Unauthorized")
 	}
+	if err := e.verifyBasicUser(user); err != nil {
+		return err
+	}
+	// if hashed password is set, validate it
+	if e.options.BasicAuthHashPassword != "" {
+		return e.validateBasicAuthHashPassword(password)
+	}
+	// else validate plain password
+	return e.validateBasicAuthPassword(password)
+}
 
+func (e *Exporter) verifyBasicUser(user string) error {
 	userCorrect := subtle.ConstantTimeCompare([]byte(user), []byte(e.options.BasicAuthUsername))
+	if userCorrect == 0 {
+		return errors.New("Unauthorized")
+	}
+	return nil
+}
+
+func (e *Exporter) validateBasicAuthHashPassword(hashedPassword string) error {
+	return bcrypt.CompareHashAndPassword([]byte(e.options.BasicAuthHashPassword), []byte(hashedPassword))
+}
+
+func (e *Exporter) validateBasicAuthPassword(password string) error {
 	passCorrect := subtle.ConstantTimeCompare([]byte(password), []byte(e.options.BasicAuthPassword))
 
-	if userCorrect == 0 || passCorrect == 0 {
+	if passCorrect == 0 {
```

Comment:
I think we should keep the same pattern of always comparing both username and password to avoid timing attacks.

I'd rewrite `verifyBasicAuth()` and check username first (store result in "userCorrect" and then the password (depending on hash or basic) and store the result in "passCorrect" - then check and return)